src/beamlime/constructors/providers.py

```diff
@@ -303,18 +240,83 @@ def split_dict_by_filter(
     The first one of returned dictionaries contains items that pass the filter.
     The other dictionary contains the rest.
     """
-    reference = {
+    filtered = {
         key: value for key, value in sequence.items() if filter_func(key, value)
     }
-    filtered = {key: value for key, value in sequence.items() if key not in reference}
-    return reference, filtered
+    rest = {key: value for key, value in sequence.items() if key not in filtered}
+    return filtered, rest
+
 
+class _ArgumentsInstanceFilter:
+    """Filter arguments by instance equality."""
+
+    def __init__(self) -> None:
+        self.arguments: dict[str, Any]
+
+    def arguments_filter(self, arg_name: str, arg_val: Any) -> bool:
+        # self.arguments.get(arg_name) is not sufficient
+        # because it will return ``None`` even if it was not used in the previous call.
+        return arg_name in self.arguments and self.arguments.get(arg_name) is arg_val
+
+    def save_arguments(self, arguments: dict) -> None:
+        self.arguments = arguments
+
+    def was_called(self) -> bool:
+        return hasattr(self, 'arguments')
+
+    def __call__(
+        self, arguments: dict[str, Any], reverse: bool = False
+    ) -> dict[str, Any]:
+        if not self.was_called():
+            self.save_arguments(arguments)
+            filtered, rest = arguments, dict()
+        else:
+            filtered, rest = split_dict_by_filter(self.arguments_filter, arguments)
+
+        return rest if reverse else filtered
+
+
+class _ArgumentsHashFilter(_ArgumentsInstanceFilter):
```

Comment:
Sorry to be a pain, but I still think it's a little weird that the `_ArgumentsHashFilter` inherits from the `_ArgumentsInstanceFilter`, as they are doing quite different things.

If you don't want to duplicate the `__call__` and `was_called` methods, then I think it would be cleaner, and make more sense conceptually, to have maybe a `_ArgumentsFilter` mixin, and then both Instance and Hash filters could inherit from that?

Reply:
Sorry I found this comment very late...! I made an abstract class `_ArgumentFilter` as you suggested and now each type of argument filter inherit the abstract class instead of having multi-depth inheritance...! Thank you!